api/api.py

```diff
@@ -290,6 +305,48 @@ def user_notifications(user):
         db.session.commit()
         return jsonify({})
 
+@api_blueprint.route('/api/users/me/achievements', methods=['GET', 'POST'])
+@user_required
+def user_achievements(user):
+    if request.method == 'GET':
+        return jsonify({'achievements': [s.to_dict() for s in user.achievements]})
+    elif request.method == 'POST':
+        for k in m.UserAchievement.REQUIRED_FIELDS:
+            if k not in request.json:
+                return jsonify({'message': f"Missing key: {k}."}), 400
+
+        try:
+            validated = m.UserAchievement.validate_dict(request.json)
+        except m.ValidationError as e:
+            return jsonify({'message': e.message}), 400
+
+        achievement = m.UserAchievement(**validated)
+        achievement.generate_key()
+        achievement.user = user
+        db.session.add(achievement)
+        db.session.commit()
+
+        return jsonify({'key': achievement.key, 'achievement': achievement.to_dict()})
+
+@api_blueprint.route('/api/users/me/achievements/<key>', methods=['PUT', 'DELETE'])
+@user_required
+def put_delete_achievement(user, key):
+    achievement = m.UserAchievement.query.filter_by(user_id=user.id, key=key).first()
+    if not achievement:
+        return jsonify({'message': "Not found."}), 404
+
+    if request.method == 'PUT':
+        try:
+            for k, v in m.UserAchievement.validate_dict(request.json).items():
+                setattr(achievement, k, v)
```

Comment:
Can `PUT` be used to delete existing items? I mean, here you get the ones provided in the request and set them to be saved, but you're not looking if there is one already saved but not present in the request that should be deleted.

Reply:
Woow, so this is indeed an issue, and the funny thing is, this code is used everywhere (for auctions/listings). In other cases, the bug is still there, but we didn't notice it, because the only field that can be "removed" is `shipping_from`, which is a string, so it becomes empty rather than `null`. But indeed, in this case we would need to set the fields to `null`.

Reply:
I think the correct solution is - the client should send `null` for fields it wants to "unset" (basically set to `null` in the database). And the API would still only modify the fields that are part of the JSON - but if they are `null`, it would set them to `null`.